File: 백준/Gold/2252. 줄 세우기/줄 세우기.py
# 줄 세우기
import sys
input = sys.stdin.readline

n, m = map(int,input().split())
child = [[] for _ in range(n+1)]
count = [0] * (n+1)
stack = []
result = []

for i in range(m):
    A, B = map(int,input().split())
    count[B] += 1
    child[A].append(B)

for i in range(1, n+1):
    if not count[i]:
        stack.append(i)

while stack:
    cur = stack.pop()
    result.append(cur)

    for i in child[cur]:
        count[i] -= 1

        if not count[i]:
            stack.append(i)

print(*result)

File: 백준/Gold/2252. 줄 세우기/줄 세우기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        ArrayList<Integer>[] child = new ArrayList[n+1];
        int[] count = new int[n+1];
        int[] result = new int[n];
        int cnt = 0;

        for (int i = 0; i <= n; i++) {
            child[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            child[A].add(B);
            count[B] += 1;
        }

        for (int i = 1; i <= n; i++) {
            if (count[i] == 0) {
                stack.push(i);
            }
        }

        while (stack.size() > 0) {
            int cur = stack.pop();
            result[cnt++] = cur;

            for (int i : child[cur]) {
                if (--count[i] == 0) {
                    stack.push(i);
                }
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(result[i]).append(" ");
        }
        System.out.println(sb);
    }
}
